utils/timer.h

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (C) 2019 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#ifndef UTILS_TIMER_H_
+#define UTILS_TIMER_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file timer.h
+ * @brief Implementation of function timer
+ */
+
+#include <pthread.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/time.h>
+#include "accelerator/errors.h"
+
+/**
+ * Initialize logger
+ */
+void timer_logger_init();
+
+/**
+ * Release logger
+ *
+ * @return
+ * - zero on success
+ * - EXIT_FAILURE on error
+ */
+int timer_logger_release();
+
+typedef struct {
+  pthread_t *thread_id;
+  timer_t *timer_id;
+} timer_args_t;
+
+/**
+ * Run a callback and blocks the current thread until timeout or callback finishes.
+ * @param[in] thread_id Pthread id for the new thread
+ * @param[in] timer Timeout to wait until callback being halted
+ * @param[in] callback The actual function being called
+ * @param[in] args The arguments to pass to the callback function
+ *
+ * @return
+ * - PTHREAD_CANCELED on time limit exceeded
+ * - others depending on the callback function
+ */
+status_t timer_join(pthread_t *thread_id, const struct itimerspec *timer, void *callback, void *args);
+
+/**
+ * Run a callback in a different thread and cancels the thread after the given timeout
+ * @param[in] thread_id Pthread id for the new thread
+ * @param[in] timer Timeout to wait until callback being halted
+ * @param[in] callback The actual function being called
+ * @param[in] args The arguments to pass to the callback function
+ */
+void timer_fork(pthread_t *thread_id, const struct itimerspec *timer, void *callback, void *args);
```

Comment:
`timer_fork` / `timer_join` is a misleading pair for naming. Instead, something like `timer_start` and `timer_stop` might be meaningful.
Reference: https://qnaplus.com/implement-periodic-timer-linux/

Reply:
To avoid naming conflicts, you can rename the functions to `ta_timer_start` and `ta_timer_stop`.

Reply:
Should I also introduce a join function since the usage in `ta_generate_address` needs to be blocked. Or should I wrap `api_generate_address` with the timer instead of `ta_generate_address` to avoid the join operation?

Note: `api_generate_address` wraps `ta_generate_address`; however, we can't put 
 only `ta_generate_address` into a new thread since the underlying code relys to the result from `ta_generate_address`. If that's the case, we should either introduce a join operation or wrap the outer function instead.

Reply:
Once `ta_timer_stop` is a synchronous function, it can accept the vector of return values by the callback functions triggered by ` ta_timer_start`.

Reply:
Alternatively, follow the idea behind CMSIS to implement one-shot timer: https://www.keil.com/pack/doc/CMSIS/RTOS2/html/group__CMSIS__RTOS__TimerMgmt.html

Reply:
`ta_timer_start` initiates the callback function, and `ta_timer_stop` to wait and join the callback function. Would it make sense?